pkg/registry/generic/etcd/etcd.go

```diff
@@ -459,16 +462,56 @@ func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, lis
 	if err != nil {
 		return nil, err
 	}
-	for _, item := range items {
-		accessor, err := meta.Accessor(item)
-		if err != nil {
-			return nil, err
-		}
-		if _, err := e.Delete(ctx, accessor.GetName(), options); err != nil {
-			return nil, err
+	// Spawn a number of goroutines, so that we can issue requests to etcd
+	// in parallel to speed up deletion.
+	workersNumber := registryoptions.GetDeleteCollectionParallelism()
```

Comment:
+1000 for options structs or arguments over globals.

On Fri, Feb 19, 2016 at 10:21 AM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/registry/generic/etcd/etcd.go
> <https://github.com/kubernetes/kubernetes/pull/21469#discussion_r53496809>
> :
>
> > @@ -459,16 +462,56 @@ func (e *Etcd) DeleteCollection(ctx api.Context, options *api.DeleteOptions, lis
> >  	if err != nil {
> >  		return nil, err
> >  	}
> > -	for _, item := range items {
> > -		accessor, err := meta.Accessor(item)
> > -		if err != nil {
> > -			return nil, err
> > -		}
> > -		if _, err := e.Delete(ctx, accessor.GetName(), options); err != nil {
> > -			return nil, err
> > +	// Spawn a number of goroutines, so that we can issue requests to etcd
> > +	// in parallel to speed up deletion.
> > +	workersNumber := registryoptions.GetDeleteCollectionParallelism()
>
> Passing arguments to things is how we do stuff. Globals are a shortcut
> that leads to worse maintenance and less testability. If you're concerned
> about more arguments, we should add a "RESTOptions" struct in pkg/api/rest
> and pass it to generic etcd across all rest helpers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21469/files#r53496809>.
>